hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java

```diff
@@ -1071,4 +1072,43 @@ private HoodieTable getTableAndInitCtx() {
     }
     return table;
   }
+
+  private void updateMetadataAndRollingStats(String actionType, HoodieCommitMetadata metadata, List<Tuple2<String,
+      HoodieWriteStat>> stats) {
+    // TODO : make sure we cannot rollback / archive last commit file
+    try {
+      // Create a Hoodie table which encapsulated the commits and files visible
+      HoodieTable table = HoodieTable.getHoodieTable(
+          new HoodieTableMetaClient(jsc.hadoopConfiguration(), config.getBasePath(), true), config, jsc);
+      // 0. All of the rolling stat management is only done by the DELTA commit for MOR and COMMIT for COW other wise
+      // there may be race conditions
+      HoodieRollingStatMetadata rollingStatMetadata = new HoodieRollingStatMetadata(actionType);
+      // 1. Look up the previous compaction/commit and get the HoodieCommitMetadata from there.
+      // 2. Now, first read the existing rolling stats and merge with the result of current metadata.
+
+      // Need to do this on every commit (delta or commit) to support COW and MOR.
+      for (Tuple2<String, HoodieWriteStat> stat : stats) {
+        metadata.addWriteStat(stat._1(), stat._2());
+        HoodieRollingStat hoodieRollingStat = new HoodieRollingStat(stat._2().getFileId(), stat._2().getNumWrites(),
+            stat._2().getNumUpdateWrites(), stat._2().getTotalWriteBytes());
+        rollingStatMetadata.addRollingStat(stat._1, hoodieRollingStat);
+      }
+      // The last rolling stat should be present in the completed timeline
+      Optional<HoodieInstant> lastInstant = table.getActiveTimeline().getDeltaCommitTimeline().filterCompletedInstants()
+          .lastInstant();
+      if (lastInstant.isPresent()) {
+        HoodieCommitMetadata commitMetadata = HoodieCommitMetadata
+            .fromBytes(table.getActiveTimeline().getInstantDetails(lastInstant
+                .get()).get());
+        rollingStatMetadata = rollingStatMetadata
```

Comment:
if a compaction runs and thus the log becomes empty now, don't we have to reset the stats? We always seem to be merging with previous stats?

Reply:
well, the stats are the same irrespective of whether it's log files or parquet files. Resetting the stats could be a good size optimization since the stats will grow over time. It is tricky to erase the stats based on compaction since we have parallel compactions running. For example, say we look at the timeline and get all the compaction commits present after the last deltacommit, then we have to look up all the fileId's for all those compaction commits, if there was no new log file written for that file id while that compaction took place, then reset the stats, else keep the stats since there is a phantom parquet for which new log files were created. There are other corner cases too which I cannot think of now (but had considered this approach when I first opened the PR). If you strongly feel we should invest in the space optimization aspect, I can spend time digging deeper into this to list out the issues. Let me know.